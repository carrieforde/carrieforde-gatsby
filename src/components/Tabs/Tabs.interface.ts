import { ReactNode } from 'react';

export interface TabsProps {
  tabItems: Tab[];
  ariaLabel: string;
}

export interface Tab {
  title: string; // what appears on the button
  content: ReactNode & string;
}

export interface TabButtonProps {
  id: string;
  ariaSelected: boolean;
  onClick: () => void;
}

export interface TabPanelProps {
  id: string;
  hidden: boolean;
}
