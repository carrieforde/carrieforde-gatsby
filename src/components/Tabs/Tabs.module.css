.tabs {
	margin-top: var(--size-6);
}

.tabsList {
	line-height: 1;
	overflow-x: auto;
	white-space: nowrap;
}

.tabsList:after {
	background-color: var(--color-neutral-2);
	content: '';
	display: block;
	height: 2px;
	width: 100%;
}

.tabsList::-webkit-scrollbar {
	width: 0;
}

.tabButton {
	font-size: var(--font-size-2);
	font-weight: 300;
	padding: 0 0 var(--size-2) 0;
	position: relative;
}

.tabButton + .tabButton {
	margin-left: var(--size-4);
}

.tabButton:after {
	background-color: var(--link-code);
	bottom: -2px;
	content: '';
	display: block;
	height: 2px;
	position: absolute;
	transition: width 0.2s ease;
	width: 0;
}

.tabButtonIsActive:after {
	width: 100%;
}

.tabPanel {
	margin-top: var(--size-5);
}

@media screen and (min-width: 600px) {
	.tabButton {
		font-size: var(--font-size-3);
	}

	.tabButton + .tabButton {
		margin-left: var(--size-6);
	}
}
