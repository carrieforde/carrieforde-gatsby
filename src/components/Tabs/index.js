import React, { Component } from 'react';
import styles from './tabs.module.css';

class Tabs extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentTab: 0
    };
  }

  render() {
    const { edges } = this.props,
      { currentTab } = this.state;

    return (
      <div className={styles.tabs}>
        <div className={styles.tabsNavigation} role="tablist" aria-label="Jobs">
          {edges.map((job, index) => {
            const { frontmatter } = job.node;

            return (
              <button
                key={index}
                id={index}
                className={
                  currentTab === index
                    ? styles.tabsButtonSelected
                    : styles.tabsButton
                }
                role="tab"
                aria-selected={currentTab === index ? true : false}
                aria-controls={`${index}-tab`}
                tabIndex="-1"
                onClick={() => this.setState({ currentTab: index })}
              >
                {frontmatter.company}
              </button>
            );
          })}
        </div>

        {edges.map((job, index) => {
          const { frontmatter, html } = job.node,
            { title, location, range, company } = frontmatter;
          return (
            <div
              key={index}
              id={`${index}-tab`}
              className={styles.tabsPanel}
              role="tabpanel"
              aria-labelledby={index}
              hidden={currentTab === index ? false : true}
              tabIndex="0"
            >
              <h3 className={styles.jobTitle}>
                {title} &#64; {company}
              </h3>
              <p className={styles.locationDate}>
                <span>{location}</span>
                <span className={styles.separator}>&#9656;</span>
                <span>{range}</span>
              </p>

              <div dangerouslySetInnerHTML={{ __html: html }} />
            </div>
          );
        })}
      </div>
    );
  }
}

export default Tabs;
