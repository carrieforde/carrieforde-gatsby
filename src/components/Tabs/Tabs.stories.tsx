import React from 'react';
import Tabs from './Tabs';
import { TabsProps } from './Tabs.interface';
import { Story, Meta } from '@storybook/react';
import Heading from 'components/Heading/Heading';

export default {
  title: 'Tabs',
  component: Tabs,
} as Meta;

const Template: Story<TabsProps> = (args) => <Tabs {...args} />;

export const Default = Template.bind({});

Default.args = {
  ariaLabel: 'Experience',
  tabItems: [
    {
      title: 'FFN',
      content: (
        <>
          <Heading level="h3">Senior Software Engineer</Heading>
          <p>I do things</p>
        </>
      ),
    },
    {
      title: 'WDS',
      content: 'Some content',
    },
    {
      title: 'MIGHTYminno',
      content: 'Some content',
    },
    {
      title: 'Cisco',
      content: 'Some content',
    },
  ],
};
