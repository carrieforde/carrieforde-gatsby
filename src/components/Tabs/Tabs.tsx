import { kebabCase } from 'lodash';
import React, { useCallback, useState } from 'react';
import { useMemo } from 'react';
import { TabButtonProps, TabPanelProps, TabsProps } from './Tabs.interface';
import {
  tabs,
  tabsList,
  tabButton,
  tabButtonIsActive,
  tabPanel,
} from './Tabs.module.css';
import cn from 'classnames';

const Tabs: React.FC<TabsProps> = ({ tabItems, ariaLabel }) => {
  const initialTab = useMemo(() => kebabCase(tabItems[0].title), []);
  const [currentTab, setCurrentTab] = useState<string>(initialTab);
  const handleClick = useCallback((key: string) => setCurrentTab(key), []);

  return (
    <div className={tabs}>
      <div className={tabsList} role="tablist" aria-label={ariaLabel}>
        {tabItems.map(({ title }) => {
          const key = kebabCase(title);
          const ariaSelected = currentTab === key ? true : false;

          return (
            <TabButton
              key={key}
              id={key}
              onClick={() => handleClick(key)}
              ariaSelected={ariaSelected}
            >
              {title}
            </TabButton>
          );
        })}
      </div>

      {tabItems.map(({ title, content }) => {
        const key = kebabCase(title);
        const hidden = currentTab === key ? false : true;

        return (
          <TabPanel key={key} id={key} hidden={hidden}>
            {content}
          </TabPanel>
        );
      })}
    </div>
  );
};

const TabButton: React.FC<TabButtonProps> = ({
  id,
  ariaSelected,
  onClick,
  children,
}) => {
  const classes = cn(tabButton, { [tabButtonIsActive]: ariaSelected });

  return (
    <button
      className={classes}
      id={`${id}-tab`}
      role="tab"
      aria-selected={ariaSelected}
      aria-controls={`${id}-panel`}
      tabIndex={-1}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

const TabPanel: React.FC<TabPanelProps> = ({ id, hidden, children }) => (
  <div
    className={tabPanel}
    id={`${id}-panel`}
    role="tabpanel"
    aria-labelledby={`${id}-tab`}
    hidden={hidden}
    tabIndex={0}
  >
    {children}
  </div>
);

export default Tabs;
