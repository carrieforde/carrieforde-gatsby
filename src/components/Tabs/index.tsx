import { MDXRenderer } from 'gatsby-plugin-mdx';
import React, { useState } from 'react';
import { getComponentKey } from 'utils/utilities';
import { TabsProps } from './Tabs.interface';
import {
  tabs,
  tabsNavigation,
  tabsButtonSelected,
  tabsButton,
  tabsPanel,
  jobTitle,
  locationDate,
  separator,
} from './tabs.module.css';

const Tabs: React.FC<TabsProps> = ({ edges }) => {
  const [currentTab, setCurrentTab] = useState(0);

  return (
    <div className={tabs}>
      <div className={tabsNavigation} role="tablist" aria-label="Jobs">
        {edges.map((job, index) => {
          const { frontmatter } = job.node;
          const buttonKey = getComponentKey(frontmatter.company, index);

          return (
            <button
              key={buttonKey}
              id={`${buttonKey}-tab`}
              className={currentTab === index ? tabsButtonSelected : tabsButton}
              role="tab"
              aria-selected={currentTab === index ? true : false}
              aria-controls={`${buttonKey}-tab`}
              tabIndex={-1}
              onClick={() => setCurrentTab(index)}
            >
              {frontmatter.company}
            </button>
          );
        })}
      </div>

      {edges.map((job, index) => {
        const { frontmatter, body } = job.node;
        const { title, location, range, company } = frontmatter;
        const panelKey = `${getComponentKey(company, index)}-panel`;
        return (
          <div
            key={panelKey}
            id={panelKey}
            className={tabsPanel}
            role="tabpanel"
            aria-labelledby={getComponentKey(company, index)}
            hidden={currentTab === index ? false : true}
            tabIndex={0}
          >
            <h3 className={jobTitle}>
              {title} &#64; {company}
            </h3>
            <p className={locationDate}>
              <span>{location}</span>
              <span className={separator}>&#9656;</span>
              <span>{range}</span>
            </p>

            <div>
              <MDXRenderer>{body}</MDXRenderer>
            </div>
          </div>
        );
      })}
    </div>
  );
};

export default Tabs;
