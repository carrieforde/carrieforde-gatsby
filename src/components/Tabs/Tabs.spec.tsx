import React from 'react';
import Tabs from './Tabs';
import { render, act, fireEvent } from '@testing-library/react';
import { Default } from './Tabs.stories';

describe('Tabs', () => {
  it('should render the Tabs component', () => {
    const { container } = render(<Default {...Default.args} />);

    expect(container).toMatchSnapshot();
  });

  it('should show the correct panel when a button is clicked', () => {
    const { container } = render(<Default {...Default.args} />);
    const buttons = container.querySelectorAll('button');
    const clickedButton = buttons[2];
    const panelId = clickedButton.getAttribute('aria-controls');

    act(() => {
      fireEvent.click(clickedButton);
    });

    const panel = container.querySelector(`#${panelId}`);

    expect(panel).not.toHaveAttribute('hidden');
  });
});
