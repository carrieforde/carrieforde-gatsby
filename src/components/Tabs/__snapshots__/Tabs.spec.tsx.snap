// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render the Tabs component 1`] = `
<div>
  <div
    class="tabs"
  >
    <div
      aria-label="Experience"
      class="tabsList"
      role="tablist"
    >
      <button
        aria-controls="ffn-panel"
        aria-selected="true"
        class="tabButton tabButtonIsActive"
        id="ffn-tab"
        role="tab"
        tabindex="-1"
      >
        FFN
      </button>
      <button
        aria-controls="wds-panel"
        aria-selected="false"
        class="tabButton"
        id="wds-tab"
        role="tab"
        tabindex="-1"
      >
        WDS
      </button>
      <button
        aria-controls="might-yminno-panel"
        aria-selected="false"
        class="tabButton"
        id="might-yminno-tab"
        role="tab"
        tabindex="-1"
      >
        MIGHTYminno
      </button>
      <button
        aria-controls="cisco-panel"
        aria-selected="false"
        class="tabButton"
        id="cisco-tab"
        role="tab"
        tabindex="-1"
      >
        Cisco
      </button>
    </div>
    <div
      aria-labelledby="ffn-tab"
      class="tabPanel"
      id="ffn-panel"
      role="tabpanel"
      tabindex="0"
    >
      <h3>
        Senior Software Engineer
      </h3>
      <p>
        I do things
      </p>
    </div>
    <div
      aria-labelledby="wds-tab"
      class="tabPanel"
      hidden=""
      id="wds-panel"
      role="tabpanel"
      tabindex="0"
    >
      Some content
    </div>
    <div
      aria-labelledby="might-yminno-tab"
      class="tabPanel"
      hidden=""
      id="might-yminno-panel"
      role="tabpanel"
      tabindex="0"
    >
      Some content
    </div>
    <div
      aria-labelledby="cisco-tab"
      class="tabPanel"
      hidden=""
      id="cisco-panel"
      role="tabpanel"
      tabindex="0"
    >
      Some content
    </div>
  </div>
</div>
`;
